pkgs/macro_service/lib/src/macro_service.dart

```diff
@@ -18,3 +25,96 @@ abstract interface class MacroService {
   /// Handles [request].
   Future<Response> handle(HostRequest request);
 }
+
+extension ProtocolExtension on Protocol {
+  bool get isJson => encoding == 'json';
+  bool get isBinary => encoding == 'binary';
+
+  /// Serializes [node] and sends it to [sink].
+  void send(void Function(Uint8List) sink, Map<String, Object?> node) {
+    if (isJson) {
+      sink(const Utf8Encoder().convert(json.encode(node)));
+      sink(_utf8Newline);
+    } else if (isBinary) {
+      // Four byte message length followed by message.
+      // TODO(davidmorgan): variable length int encoding probably makes more
+      // sense than fixed four bytes.
+      final binary = serialization.serializeToBinary(node);
+      final length = binary.length;
+      sink(Uint8List.fromList([
+        length & 255,
+        (length >> 8) & 255,
+        (length >> 16) & 255,
+        (length >> 24) & 255,
+      ]));
+      sink(binary);
+    } else {
+      throw StateError('Unsupported protocol: $this.');
+    }
+  }
+
+  /// Deserializes [stream] to JSON objects.
+  ///
+  /// The data on `stream` can be arbitrarily split to `Uint8List` instances,
+  /// it does not have to be one list per message.
+  Stream<Map<String, Object?>> decode(Stream<Uint8List> stream) {
+    if (isJson) {
+      return const Utf8Decoder()
+          .bind(stream)
+          .transform(const LineSplitter())
+          .map((line) => json.decode(line) as Map<String, Object?>);
+    } else if (isBinary) {
+      return stream.transform(_BinaryDeserializer());
+    } else {
+      throw StateError('Unsupported protocol: $this');
+    }
+  }
+}
+
+/// Splits the binary format to messages and deserializes them.
+///
+/// Messages are a four byte little-endian length followed by the message data
+/// of that length.
+class _BinaryDeserializer
```

Comment:
See https://github.com/dart-lang/sdk/blob/main/pkg/_macros/lib/src/executor/message_grouper.dart. I would just copy that implementation.

We still do end up copying bytes but it uses all Uint8Lists allocated with the correct size and sets ranges within that which is a bit more optimized.

To avoid a copy we would need a buffer which can read from a list of Uint8List (views), which we could also do to be sure. But, then it won't be a Uint8List and as a result might miss out on some other optimizations, so we would need to benchmark.

Reply:
Whoops, should have realized there would be an implementation already. Will swap it in + address other comments tomorrow. Thanks :)